face.go

```diff
@@ -1,3 +1,102 @@
+	Upem() uint16
+
+	// GlyphName returns the name of the given glyph, or an empty
+	// string if the glyph is invalid or has no name.
+	GlyphName(glyph GID) string
+
+	// FontHExtents returns the extents of the font for horizontal text, or false
+	// it not available, in font units.
+	// `varCoords` (in normalized coordinates) is only useful for variable fonts.
+	FontHExtents(varCoords []float32) (FontExtents, bool)
+
+	// FontVExtents is the same as `FontHExtents`, but for vertical text.
+	FontVExtents(varCoords []float32) (FontExtents, bool)
+
+	// NominalGlyph returns the glyph used to represent the given rune,
+	// or false if not found.
+	NominalGlyph(ch rune) (GID, bool)
+
+	// HorizontalAdvance returns the horizontal advance in font units.
+	// When no data is available but the glyph index is valid, this method
+	// should return a default value (the upem number for example).
+	// If the glyph is invalid it should return 0.
+	// `varCoords` is used by variable fonts, and is specified in normalized coordinates.
+	HorizontalAdvance(glyph GID, varCoords []float32) float32
+
+	// VerticalAdvance is the same as `HorizontalAdvance`, but for vertical advance.
+	VerticalAdvance(glyph GID, varCoords []float32) float32
+
+	// GlyphHOrigin fetches the (X,Y) coordinates of the origin (in font units) for a glyph ID,
+	// for horizontal text segments.
+	// Returns `false` if not available.
+	GlyphHOrigin(glyph GID, varCoords []float32) (x, y int32, found bool)
+
+	// GlyphVOrigin is the same as `GlyphHOrigin`, but for vertical text segments.
+	GlyphVOrigin(glyph GID, varCoords []float32) (x, y int32, found bool)
+
+	// GlyphExtents retrieve the extents for a specified glyph, of false, if not available.
+	// `varCoords` is used by variable fonts, and is specified in normalized coordinates.
+	// For bitmap glyphs, the closest resolution to `xPpem` and `yPpem` is selected.
+	GlyphExtents(glyph GID, varCoords []float32, xPpem, yPpem uint16) (GlyphExtents, bool)
+
+	// NormalizeVariations should normalize the given design-space coordinates. The minimum and maximum
+	// values for the axis are mapped to the interval [-1,1], with the default
+	// axis value mapped to 0.
+	// This should be a no-op for non-variable fonts.
+	NormalizeVariations(varCoords []float32) []float32
+}
+
+// FaceRenderer implements the drawing of glyphs.
+type FaceRenderer interface {
+	// TODO: add position information and output/destination
+	Draw(glyph GID)
+}
+
+// GID is used to identify glyphs in a font.
+// It is mostly internal to the font and should not be confused with
+// Unicode code points.
+type GID uint32
+
+// FontExtents exposes font-wide extent values, measured in font units.
+// Note that typically ascender is positive and descender negative in coordinate systems that grow up.
+type FontExtents struct {
+	Ascender  float32 // Typographic ascender.
+	Descender float32 // Typographic descender.
+	LineGap   float32 // Suggested line spacing gap.
```

Comment:
just a drive-by comment: why `float32`?
shouldn't we use some fixed-point integer type (like [`x/image/math/fixed.Int26_6`](https://godocs.io/golang.org/x/image/math/fixed#Int26_6)) instead ?

Reply:
Good point, thanks ! I've used `float32` for simplicity, but `x/image/math/fixed.Int26_6` will probably be required for integration and performances.

Reply:
perhaps this should be reflected in this PR? :)

Reply:
Actually, thinking of it a bit more, it is not obvious to me that `x/image/math/fixed.Int26_6` is a better choice. From what I've read, the performance of `float32` is similar on modern computers. I'm not sure about the integration point.
I've opened [this issue](https://github.com/go-text/shaping/issues/3) to discuss the question.